### PR TITLE
API/BUG: make .at raise same exceptions as .loc

### LINKED ISSUE #31722: API/BUG: Inconsistent errors/msgs between loc vs at
```
df = DataFrame({"A": [1, 2, 3]}, index=list("abc"))

>>> df.at[0]
ValueError: At based indexing on an non-integer index can only have non-integer indexers
>>> df.loc[0]
TypeError: cannot do label indexing on <class 'pandas.core.indexes.base.Index'> with these indexers [0] of <class 'int'>
```

I would expect these to behave the same.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #31722
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This also (very) indirectly addresses #31683 which in turn will let us get rid of `CategoricalIndex.get_value` altogether.

### CODE REVIEW

File: pandas/tests/indexing/test_scalar.py
Comment (WillAyd) [inline pandas/tests/indexing/test_scalar.py]: Can you parametrize this test for at / loc? I think would make it easier and more comprehensive

File: doc/source/whatsnew/v1.1.0.rst
Comment (WillAyd) [inline doc/source/whatsnew/v1.1.0.rst]: Should mention the `KeyError` here as well?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/indexing/test_scalar.py
Comment (jbrockmendel) [reply pandas/tests/indexing/test_scalar.py]: looking at this, im definitely going to split this into more targeted tests, will look at parametrizing; might benefit from a helper to assert we get the same behavior

Comment (jbrockmendel) [reply pandas/tests/indexing/test_scalar.py]: refactored to make targeted tests, this file needs a dedicated pass to move old testing patterns to parametrization

File: pandas/tests/indexing/test_scalar.py
Comment (jorisvandenbossche) [inline pandas/tests/indexing/test_scalar.py]: I have to say that I found the previous error message more readable ..

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
Can you check some timings for `at`? (as this is supposed to be the fast one)

How does this indirectly also address #31683 ? And if it does, shouldn't there be tests involving categorical / a whatsnew?

**Comment by jbrockmendel** [comment]
> Can you check some timings for at? (as this is supposed to be the fast one)

Yah we take a hit

```
In [2]: ser = pd.Series(range(10000))                                           
In [3]: %timeit ser.at[501]                                                     
4.54 µs ± 139 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- master
7.8 µs ± 75.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- PR

In [4]: df = ser.to_frame("A")                                                  
In [5]: %timeit df.at[501, "A"]                                                 
3.9 µs ± 25.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- master
7.71 µs ± 99.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- PR
```

We might get some of that back from #31709.

File: pandas/tests/indexing/test_scalar.py
Comment (jbrockmendel) [reply pandas/tests/indexing/test_scalar.py]: I generally agree, that suggests we should improve the existing messages for "loc"

**Comment by jbrockmendel** [comment]
> How does this indirectly also address #31683 ? And if it does, shouldn't there be tests involving categorical / a whatsnew?

With this in place, all the places from which index.get_value are called are preceeded by maybe_cast_scalar_indexer (except one in which the key is a tuple).  That makes the call inside CategoricalIndex.get_value unnecessary, which addresses the motivating issue for #31683.  The "indexing on CategoricalIndex is hard" problem is still a problem.

**Review by jreback — APPROVED** [review]
consistency is a huge + here, makes the code *much* better.

I am not sure if @WillAyd or @jorisvandenbossche have comments, but lgtm.

**Review by WillAyd — APPROVED** [review]
lgtm. parametrization will be a nice follow on

**Comment by jreback** [comment]
thanks @jbrockmendel

**Comment by jorisvandenbossche** [comment]
@jreback maybe you didn't notice, but *this time* I had a "request changes" open, as you asked previous time to indicate that you shouldn't merge without me having the chance to answer back. 
If there is still discussion going on, please give it some time.

**Comment by jorisvandenbossche** [comment]
@jbrockmendel can you do a follow up with:

- fixing the performance regression
- fixing the usability regression in error message

**Comment by jorisvandenbossche** [comment]
> fixing the usability regression in the error message

Looking back at it, it's actually mainly the long fully qualified class name that is annoying, so that is an easy fix -> https://github.com/pandas-dev/pandas/pull/31769

**Pull request merged**
